src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatCompletion.cs

```diff
@@ -0,0 +1,616 @@
+                        result.Tools.Add(ToOpenAIChatTool(af));
+                    }
+                }
+
+                switch (options.ToolMode)
+                {
+                    case AutoChatToolMode:
+                        result.ToolChoice = ChatToolChoice.CreateAutoChoice();
+                        break;
+
+                    case RequiredChatToolMode required:
+                        result.ToolChoice = required.RequiredFunctionName is null ?
+                            ChatToolChoice.CreateRequiredChoice() :
+                            ChatToolChoice.CreateFunctionChoice(required.RequiredFunctionName);
+                        break;
+                }
+            }
+
+            if (options.ResponseFormat is ChatResponseFormatText)
+            {
+                result.ResponseFormat = OpenAI.Chat.ChatResponseFormat.CreateTextFormat();
+            }
+            else if (options.ResponseFormat is ChatResponseFormatJson jsonFormat)
+            {
+                result.ResponseFormat = jsonFormat.Schema is { } jsonSchema ?
+                    OpenAI.Chat.ChatResponseFormat.CreateJsonSchemaFormat(
+                        jsonFormat.SchemaName ?? "json_schema",
+                        BinaryData.FromBytes(
+                            JsonSerializer.SerializeToUtf8Bytes(jsonSchema, OpenAIJsonContext.Default.JsonElement)),
+                        jsonFormat.SchemaDescription) :
+                    OpenAI.Chat.ChatResponseFormat.CreateJsonObjectFormat();
+            }
+        }
+
+        return result;
+    }
+
+    private static AITool FromOpenAIChatTool(ChatTool chatTool)
+    {
+        Dictionary<string, object?> additionalProperties = new();
+        if (chatTool.FunctionSchemaIsStrict is bool strictValue)
+        {
+            additionalProperties["Strict"] = strictValue;
+        }
+
+        OpenAIChatToolJson openAiChatTool = JsonSerializer.Deserialize(chatTool.FunctionParameters.ToMemory().Span, OpenAIJsonContext.Default.OpenAIChatToolJson)!;
+        List<AIFunctionParameterMetadata> parameters = new(openAiChatTool.Properties.Count);
+        foreach (KeyValuePair<string, JsonElement> property in openAiChatTool.Properties)
+        {
+            parameters.Add(new(property.Key)
+            {
+                Schema = property.Value,
+                IsRequired = openAiChatTool.Required.Contains(property.Key),
+            });
+        }
+
+        AIFunctionMetadata metadata = new(chatTool.FunctionName)
+        {
+            Description = chatTool.FunctionDescription,
+            AdditionalProperties = additionalProperties,
+            Parameters = parameters,
+            ReturnParameter = new()
+            {
+                Description = "Return parameter",
+                Schema = _defaultParameterSchema,
+            }
+        };
+
+        return new MetadataOnlyAIFunction(metadata);
+    }
+
+    private sealed class MetadataOnlyAIFunction(AIFunctionMetadata metadata) : AIFunction
```

Comment:
Separate from this PR, I wonder if we should expose an `AIFunction.Create` that just takes the metadata and a delegate for the InvokeCoreAsync implementation. Pro, it'd avoid needing to define a class like this for anyone who needs it. Con, it might be confusing with AIFunctionFactory.Create.

Reply:
Sounds reasonable, although I suspect it would require a separate method group.